### PR TITLE
Fixes #12 and #30: Add ability to change the app title

### LINKED ISSUE #12: Support changing app title
The app title says '⚡️electric io' right now but who really cares if it doesn't stay that way.
It would be cool to let people change it to say whatever they want!

+ [ ] create a new property called `title` in the dashboard.json settings
+ [ ] set the default value of the property in the json file to be `⚡️electric io` (check lib/helpers.js)
+ [ ] create a new text input field in the dashboard settings form for changing the title (follow the strategy for the background color field)
+ [ ] ensure that the title setting is saved correctly to the dashboard.json file when the 'save' button is clicked in the dashboard settings interface

### LINKED ISSUE #30: Detect emoji in title and add styling accordingly
The [PR](https://github.com/noopkat/electric-io/pull/29) to add support for app title changes removed the styling for the default emoji. Support should be added to detect emoji in the title and appropriately add the styling to them.

### PR DESCRIPTION
**Pull request by mpaarating**
Add ability to change the app title via the app settings widget.

Refactor the main app component to have the correct class applied to the
title emoji. As I'm not that good at regex (I googled this regex), it
has a capture group for emoji at the beginning of the title. It'd be
awesome if someone actually good at regex could improve this to detect
multiple emoji and emoji in different positions of the title.

Closes #12, #30.

### CODE REVIEW

**Comment by noopkat** [comment]
Hi @mpaarating,

Thanks for the hard work on this feature - I appreciate the thoughtfulness about the emoji detection!
I'll stress test this regex before merging but it should be fine. 

One thing we can possibly look into is using a greedy search for emojis and replacing them with the span wrapped version. Let me know if you can think of any issues we might run into with that approach.

I should be able to merge this soon! Thanks again 💯

**Comment by mpaarating** [comment]
I totally agree about the greedy search, I just couldn't get it to work with the time I had put in. It shouldn't be terribly difficult adjust though!

**Comment by noopkat** [comment]
@mpaarating good point! Also I never would have known you were new to Vue everything looks great. I am pretty new to Vue also (first project!) so we have that in common 😄

**Comment by mpaarating** [comment]
@noopkat I watched the stream where you started to review this PR and found some issues with it. If you don't mind, I'll steal what you did on the stream and fix the outstanding issue with the input field not showing the default title.

**Comment by noopkat** [comment]
That works for me, thanks!

File: public/js/components/app.js [hunk 1]
```diff
@@ -35,73 +40,97 @@ const initialData = function() {
...
+      var appTitle =
+        typeof this.dashboard.title === 'undefined'
+          ? '\u26a1electric io'
+          : this.dashboard.title;
+
+      var title = TITLE_EMOJI_REGEX.exec(appTitle);
+
+      if (title) {
+        title = `<span class="hemoji">${title[1]}</span>${title[2]}`;
+        this.dashboard.title = appTitle;
```
[the remark this answers is not in the document]
Comment (noopkat) [reply public/js/components/app.js]: hey there. I appreciate you working on this to implement it.
Ultimately I agree with you about this being gross, and I think that if it's not too much trouble we should drop it entirely. I think it's better moving forward if the code is less complex.
This dashboard is still a work in progress so there'll be small nits like this when new features are introduced. It's fine that they have to enter their own title the first time if they used the dashboard before this release. Thoughts? Sorry to trouble you!

**Comment by mpaarating** [comment]
Yes that sounds perfect!

File: public/js/components/app.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpaarating) [reply public/js/components/app.js]: @noopkat Do you mean drop this PR altogether or just the oddity with the title changing when you have a dashboard without the title property?

Alternatively, I was thinking about having a helper function look for the title prop in the json file and add it if it doesn't exist (on startup).

[the remark this answers is not in the document]
Comment (noopkat) [reply public/js/components/app.js]: I meant to drop the oddity with the title changing when you don't have the title property. I definitely want your work merged in as it's a needed feature!

[the remark this answers is not in the document]
Comment (noopkat) [reply public/js/components/app.js]: > Alternatively, I was thinking about having a helper function look for the title prop in the json file and add it if it doesn't exist 

This is a great idea, but let's keep this small edge case as it is and let people set the title for the first time if they want to :)

**Comment by mpaarating** [comment]
@noopkat I just reverted the commit with the BAD THINGS in it. We should be good to go otherwise! Thanks for all the feedback!

**Comment by noopkat** [comment]
@mpaarating thank so much for your patience, things got a bit busy for me with my schedule now that I'm traveling again 🙇‍♀️ 

Love the work you did on this pull request. Thank you so much!

bitmoji

**Pull request merged**